contrib/check-semver-feature.sh

```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/env bash
+#
+# Checks semver compatibility between the `--no-features` and `all-features`.
+# This is important since it tests for the presence non-additive cargo features.
+#
+# Under the hood uses cargo semver-checks to check for breaking changes.
+# We cannot use it directly since it only supports checking against published
+# crates.
+# That's the intended use case for cargo semver-checks:
+# you run before publishing a new version of a crate to check semver breaks.
+# Here we are hacking it by first generating JSON files from cargo doc
+# and then using those files to check for breaking changes with
+# cargo semver-checks.
+
+set -euo pipefail
+
+# These are the hardcoded flags that cargo semver-checks uses
+# under the hood to invoke rustdoc.
+RUSTDOCFLAGS="-Z unstable-options --document-private-items --document-hidden-items --output-format=json --cap-lints=allow"
+
+main() {
+    # Generate JSON files for no-features and all-features
+    # 1. bitcoin
+    generate_json_files_all_features "bitcoin"
+    generate_json_files_no_default_features "bitcoin"
+
+    # 2. base58ck
+    generate_json_files_all_features "base58ck"
+    generate_json_files_no_default_features "base58ck"
+
+    # 3. bitcoin_hashes
+    generate_json_files_all_features "bitcoin_hashes"
+    generate_json_files_no_default_features "bitcoin_hashes"
+
+    # 4. bitcoin-units
+    generate_json_files_all_features "bitcoin-units"
+    generate_json_files_no_default_features "bitcoin-units"
+
+    # 5. bitcoin-io
+    generate_json_files_all_features "bitcoin-io"
+    generate_json_files_no_default_features "bitcoin-io"
+
+    # Check for API semver non-addivite cargo features on all the generated JSON files above.
+    run_cargo_semver_check "bitcoin"
+    run_cargo_semver_check "base58ck"
+    run_cargo_semver_check "bitcoin_hashes"
+    run_cargo_semver_check "bitcoin-units"
+    run_cargo_semver_check "bitcoin-io"
+
+    # Invoke cargo semver-checks to check for non-additive cargo features
+    # in all generated files.
+    check_for_non_addivite_cargo_features
+}
+
+# Run cargo doc with the cargo semver-checks rustdoc flags.
+# We don't care about dependencies.
+run_cargo_doc() {
+    RUSTDOCFLAGS="$RUSTDOCFLAGS" RUSTC_BOOTSTRAP=1 cargo doc --no-deps "$@"
+}
+
+# Run cargo semver-check
+run_cargo_semver_check() {
+    local crate="$1"
+
+    echo "Running cargo semver-checks for $crate"
+    # Hack to not fail on errors.
+    # This is necessary since cargo semver-checks will fail if the
+    # semver check fails.
+    # We check that manually later.
+    set +e
+    cargo semver-checks -v --baseline-rustdoc "$crate-no-default-features.json" --current-rustdoc "$crate-all-features.json" > "$crate--semver.txt" 2>&1
```

Comment:
Consider changing the file name pattern to be different from the one used in the breaking change checking script, so that if the two steps end up running inside the same job runner, you don't get confusing errors.
```suggestion
    cargo semver-checks -v --baseline-rustdoc "$crate-no-default-features.json" --current-rustdoc "$crate-all-features.json" > "$crate--additive-features.txt" 2>&1
```